examples/edwards25519-hash/src/edwards25519-hash.rs

```diff
@@ -0,0 +1,534 @@
+//
+// The crate has 3 different specifications of the map_to_curve primitive. The
+// first two are map_to_curve_elligator2 and map_to_curve_elligator2_straight.
+// Both of these specifications utilize the monty_to_edw primitive. The third
+// specification is map_to_curve_elligator2_edwards25519. An equality test of
+// these 3 implementations and an external one can be seen in 
+// ../tests/test_equiv.rs. 
+
+use hacspec_lib::*;
+use hacspec_edwards25519::*;
+use hacspec_sha512::*;
+
+const B_IN_BYTES: usize = 64usize;
+const S_IN_BYTES: usize = 128usize;
+const L: usize = 48usize;
+const J: u128 = 486662u128;
+const Z: u128 = 2u128;
+
+array!(ArrEd25519FieldElement, 4, U64);
+
+public_nat_mod!(
+    type_name: EdFieldHash,
+    type_of_canvas: EdFieldHashCanvas,
+    bit_size_of_field: 512,
+    modulo_value: "7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffed"
+);
+
+// (p - 1) / 2
+const P_1_2: ArrEd25519FieldElement = ArrEd25519FieldElement(secret_array!(
+    U64,
+    [
+        0x3fffffffffffffffu64,
+        0xffffffffffffffffu64,
+        0xffffffffffffffffu64,
+        0xfffffffffffffff6u64
+    ]
+));
+
+// (p + 3) / 8
+const P_3_8: ArrEd25519FieldElement = ArrEd25519FieldElement(secret_array!(
+    U64,
+    [
+        0x0fffffffffffffffu64,
+        0xffffffffffffffffu64,
+        0xffffffffffffffffu64,
+        0xfffffffffffffffeu64
+    ]
+));
+
+// (p - 5) / 8
+const P_5_8: ArrEd25519FieldElement = ArrEd25519FieldElement(secret_array!(
+    U64,
+    [
+        0x0fffffffffffffffu64,
+        0xffffffffffffffffu64,
+        0xffffffffffffffffu64,
+        0xfffffffffffffffdu64
+    ]
+));
+
+// Taken from bls12-381-hash.rs
+// Specification can be seen in section 5.3.1 of https://datatracker.ietf.org/doc/draft-irtf-cfrg-hash-to-curve/
+pub fn expand_message_xmd(
+    msg: &ByteSeq, dst: &ByteSeq, len_in_bytes: usize
+) -> ByteSeq {
+    let ell = (len_in_bytes + B_IN_BYTES - 1) / B_IN_BYTES; // ceil(len_in_bytes / b_in_bytes)
+                                                            // must be that ell <= 255
+    let dst_prime = dst.push(&U8_from_usize(dst.len())); // DST || I2OSP(len(DST), 1)
+    let z_pad = ByteSeq::new(S_IN_BYTES); // I2OSP(0, s_in_bytes)
+    let mut l_i_b_str = ByteSeq::new(2);
+    l_i_b_str[0] = U8_from_usize(len_in_bytes / 256);
+    l_i_b_str[1] = U8_from_usize(len_in_bytes); // I2OSP(len_in_bytes, 2)
+
+    let msg_prime = z_pad
+        .concat(msg)
+        .concat(&l_i_b_str)
+        .concat(&ByteSeq::new(1))
+        .concat(&dst_prime); // Z_pad || msg || l_i_b_str || 0 || dst_prime
+    
+    let b_0 = ByteSeq::from_seq(&hash(&msg_prime)); // H(msg_prime)
+    let mut b_i = ByteSeq::from_seq(&hash(&b_0.push(&U8(1u8))
+        .concat(&dst_prime))); // H(b_0 || 1 || dst_prime)
+    let mut uniform_bytes = ByteSeq::from_seq(&b_i);
+
+    for i in 2..(ell + 1) {
+        let t = ByteSeq::from_seq(&b_0);
+        b_i = ByteSeq::from_seq(&hash(&(t ^ b_i).push(&U8_from_usize(i))
```

Comment:
`U8_from_usize` doesn't actually check that the input doesn't overflow.
Looks like this function should do the check manually and return a `Result`.
`len_in_bytes` should be checked as well.